src/ellipse.rs

```diff
@@ -0,0 +1,132 @@
+//! Implementation of ellipse shape.
+
+use std::f64::consts::PI;
+use std::{
+    iter,
+    ops::{Add, Sub},
+};
+
+use crate::{Arc, ArcAppendIter, PathEl, Point, Rect, Shape, Vec2};
+
+/// A ellipse.
+#[derive(Clone, Copy, Default, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Ellipse {
+    /// The center.
+    pub center: Point,
+    /// The radii.
+    pub radii: Vec2,
+    /// How much to rotate the major/minor axes.
+    pub x_rotation: f64,
+}
+
+impl Ellipse {
+    /// A new ellipse from center and radii.
+    #[inline]
+    pub fn new(center: impl Into<Point>, radii: impl Into<Vec2>, x_rotation: f64) -> Ellipse {
+        Ellipse {
+            center: center.into(),
+            radii: radii.into(),
+            x_rotation,
+        }
+    }
+}
+
+impl Add<Vec2> for Ellipse {
+    type Output = Ellipse;
+
+    #[inline]
+    fn add(self, v: Vec2) -> Ellipse {
+        Ellipse {
+            center: self.center + v,
+            radii: self.radii,
+            x_rotation: self.x_rotation,
+        }
+    }
+}
+
+impl Sub<Vec2> for Ellipse {
+    type Output = Ellipse;
+
+    #[inline]
+    fn sub(self, v: Vec2) -> Ellipse {
+        Ellipse {
+            center: self.center - v,
+            radii: self.radii,
+            x_rotation: self.x_rotation,
+        }
+    }
+}
+
+impl Shape for Ellipse {
+    type BezPathIter = iter::Chain<iter::Once<PathEl>, ArcAppendIter>;
+
+    fn to_bez_path(&self, tolerance: f64) -> Self::BezPathIter {
+        Arc {
+            center: self.center,
+            radii: self.radii,
+            start_angle: 0.0,
+            sweep_angle: 2.0 * PI,
+            x_rotation: self.x_rotation,
+        }
+        .to_bez_path(tolerance)
+    }
+
+    #[inline]
+    fn area(&self) -> f64 {
+        PI * self.radii.x * self.radii.y
+    }
+
+    #[inline]
+    fn perimeter(&self, accuracy: f64) -> f64 {
+        self.clone()
+            .into_bez_path(0.1)
+            .elements()
+            .perimeter(accuracy)
+    }
+
+    fn winding(&self, pt: Point) -> i32 {
+        self.clone().into_bez_path(0.1).elements().winding(pt)
+    }
+
+    #[inline]
+    fn bounding_box(&self) -> Rect {
+        self.clone().into_bez_path(0.1).elements().bounding_box()
+    }
+}
+
+/*
+#[cfg(test)]
```

Comment:
any reason not to include this?

Reply:
It's just a cut and past from circle atm. It should either be updated to use ellipses (preferable) ore be deleted before this PR is accepted.